### PR TITLE
DHP-1071 CSV Storage API

### PR DESCRIPTION
**Pull request by DwayneJengSage**
https://sagebionetworks.jira.com/browse/DHP-1071

API and object model for storing table rows. This is written by the Exporter3Worker to store table rows. This is read by the (currently non-existent) CSV Worker to assemble the rows into a CSV.

TODO: unit tests

### CODE REVIEW

File: src/main/resources/db/changelog/changelog.sql [hunk 1]
```diff
@@ -1102,3 +1102,25 @@ ADD COLUMN `postProcessingStatus` varchar(255) DEFAULT NULL;
...
 ADD COLUMN `phase` enum('DRAFT', 'REVIEW', 'PUBLISHED') DEFAULT 'DRAFT';
+
+-- changeset bridge:82
+
+CREATE TABLE IF NOT EXISTS `UploadTableRows` (
+    `appId` varchar(60) NOT NULL,
+    `studyId` varchar(60) NOT NULL,
+    `recordId` char(24) NOT NULL,
+    `assessmentId` varchar(255) NOT NULL,
+    `assessmentRevision` int(10) unsigned NOT NULL,
```
Comment (DwayneJengSage) [inline src/main/resources/db/changelog/changelog.sql]: Change to assessmentGuid

File: src/main/java/org/sagebionetworks/bridge/spring/controllers/UploadTableController.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        this.uploadTableService = uploadTableService;
+    }
+
+    /** Delete a single upload table row. This is currently only used by integration tests (so only superadmin). */
+    @DeleteMapping("/v1/apps/{appId}/studies/{studyId}/uploadtable/{recordId}")
+    public StatusMessage deleteUploadTableRowForSuperadmin(@PathVariable String appId, @PathVariable String studyId,
+            @PathVariable String recordId) {
+        getAuthenticatedSession(SUPERADMIN);
+
+        // Verify study exists.
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/spring/controllers/UploadTableController.java]: Add comment throwsException=true

File: src/main/java/org/sagebionetworks/bridge/spring/controllers/UploadTableController.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        studyService.getStudy(appId, studyId, true);
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/spring/controllers/UploadTableController.java]: Move this to the service.

File: src/main/java/org/sagebionetworks/bridge/spring/controllers/UploadTableController.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    @GetMapping("/v1/apps/{appId}/studies/{studyId}/uploadtable/{recordId}")
+    public UploadTableRow getUploadTableRowForSuperadmin(@PathVariable String appId, @PathVariable String studyId,
+            @PathVariable String recordId) {
+        getAuthenticatedSession(SUPERADMIN);
+
+        // Verify study exists.
+        studyService.getStudy(appId, studyId, true);
+
+        return uploadTableService.getUploadTableRow(appId, studyId, recordId).orElseThrow(
+                () -> new EntityNotFoundException(UploadTableRow.class));
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/spring/controllers/UploadTableController.java]: Move orElseThrow to the service.

File: src/main/resources/db/changelog/changelog.sql (hunk 1 above)
Comment (syoung-smallwisdom) [reply src/main/resources/db/changelog/changelog.sql]: I believe the assessment guid includes the assessment revision.

**Review by syoung-smallwisdom — APPROVED** [review]
Looks good.

I think you are going to need a second table to store the json type and description for each key-value pair in both the `data` and `metadata`. That's the first thing every researcher always asks for (which is how we ended up with the requirement of including json schemas).

**Comment by DwayneJengSage** [comment]
@syoung-smallwisdom 

> I think you are going to need a second table to store the json type and description for each key-value pair in both the `data` and `metadata`. That's the first thing every researcher always asks for (which is how we ended up with the requirement of including json schemas).

This is documention the documentation that I intended to store it in AssessmentResources as part of https://sagebionetworks.jira.com/browse/DHP-1070

Last I checked, we decided not to do that, and I don't think anyone's proposed an alternative yet 🤷

**Comment by syoung-smallwisdom** [comment]
> @syoung-smallwisdom
> 
> > I think you are going to need a second table to store the json type and description for each key-value pair in both the `data` and `metadata`. That's the first thing every researcher always asks for (which is how we ended up with the requirement of including json schemas).
> 
> This is documention the documentation that I intended to store it in AssessmentResources as part of https://sagebionetworks.jira.com/browse/DHP-1070
> 
> Last I checked, we decided not to do that, and I don't think anyone's proposed an alternative yet 🤷

I'm pointing it out as a warning. Storing all JSON values as strings loses type information that is needed to restore the value. Since strings aren't inherently typed for int/double but most serializers are, this leads to errors. Also, we inevitably get asked "what do the values _mean_?" where what they want to know is "what was the question asked?".

In the past, we've handled this by linking to schemas defined using [json-schema.org](https://json-schema.org) schemas. I would recommend a JIRA to add a table for storing the mapping of `[key : JsonSchemaProperty]` for each assessment.

While in actuality, [JsonSchemaProperty](https://github.com/BridgeDigitalHealth/JsonModel-Swift/blob/main/Sources/JsonModel/JsonSchema.swift) is somewhat obfuscated and difficult to describe in Java or Kotlin, it could be simplified to:

```
enum class JsonType { string, integer, number, boolean, array, object }

data class JsonSchemaProperty(
    type: JsonType, 
    description: String,

    // The schema for the items in an array.
    items: ArrayItems? = null,
)

data class ArrayItems(type: JsonType, enum: List<String>? = null)
``` 

This is about the limit of what most researchers can seem to grok. Attempts to give researchers more than this have led to political intrigue, but not giving them this has lead to "This is useless and doesn't tell me what I need to know".

Further reading: [Json Schema](https://cswr.github.io/JsonSchema/)

Note: I'm describing this in Kotlin b/c that's what we're planning to use for surveys and for those, this would need to be derived from the config, but I'm using Json-schema syntax b/c that's more widely used.

**Comment by DwayneJengSage** [comment]
@syoung-smallwisdom Yes, I agree that this needs to be done, but the original plan was to solve it with Assessment Resources, and it’s currently tracked by https://sagebionetworks.jira.com/browse/DHP-1070

But we decided not to do that, and I don’t know why, so I don’t know what the plan is, and I don’t know how to move forward with this 🤷

**Review by nategbrown9 — APPROVED** [review]
Looks good.

**Pull request merged**